keps/sig-scheduling/5075-dra-consumable-capacity/README.md

```diff
@@ -0,0 +1,1361 @@
+   //
+   // +required
+   Default resource.Quantity `json:"default,omitempty" protobuf:"bytes,1,name=default"`
+
+
+   // Options defines a discrete set of additional valid capacity values that can be requested.
+   // +optional
+   // +listType=atomic
+   Options []resource.Quantity `json:"options,omitempty" protobuf:"bytes,2,rep,name=options"`
+}
+
+
+// CapacityClaimPolicyRange defines a valid range of consuming capacity.
+type CapacityClaimPolicyRange struct {
+   // Minimum specifies the minimum capacity required for a consumption request.
+   // This field is required and serves as the default capacity to be consumed.
+   //
+   // +required
+   Minimum resource.Quantity `json:"minimum,omitempty" protobuf:"bytes,1,name=minimum"`
+
+
+   // Maximum specifies the maximum capacity that can be requested in a consumption request.
+   //
+   // +optional
+   Maximum *resource.Quantity `json:"maximum,omitempty" protobuf:"bytes,2,opt,name=maximum"`
+
+
+   // Step defines the incremental block size by which capacity can increase from the minimum.
+   //
+   // +optional
+   Step *resource.Quantity `json:"step,omitempty" protobuf:"bytes,3,opt,name=step"`
+}
+
+```
+
+#### ResourceClaimSpec's DeviceRequest
+
+```go
+
+type DeviceRequest struct {
+...
+   // Capacities define resource requirements against each capacity.
+   //
+   // +optional
+   // +featureGate=DRAConsumableCapacity
+   Capacities *CapacityRequirements `json:"capacities,omitempty" protobuf:"bytes,9,opt,name=capacities"`
+}
+
+
+// CapacityRequirements defines the capacity requirements for a specific device request.
+type CapacityRequirements struct {
+   // Requests specifies the requested quantity of device capacities for a device request.
+   // +optional
+   Requests map[QualifiedName]resource.Quantity `json:"requests,omitempty" protobuf:"bytes,1,rep,name=requests"`
+
+
+   // Potentially enhancement field.
+   // Limits define the maximum amount of per-device resources allowed.
+   // This enables burstable usage when applicable.
+}
+
+
+// DeviceConstraint must have exactly one field set besides Requests.
+type DeviceConstraint struct {
+	...
+   // DistinctAttribute requires that all devices in question have this
+   // attribute and that its type and value are unique across those devices.
+   //
+   // For example, specify attribute name to get virtual devices from distinct shared physical devices.
+   //
+   // +optional
+   // +oneOf=ConstraintType
+   DistinctAttribute *FullyQualifiedName `json:"distinctAttribute,omitempty" protobuf:"bytes,3,opt,name=distinctAttribute"`
+}
+
+```
+
+#### ResourceClaimStatus's DeviceRequestAllocationResult
+
+```go
+
+type DeviceRequestAllocationResult struct {
+ ...
+
+
+   // Shared indicates whether the allocated device can be shared by multiple claims.
+   //
+   // +optional
+   // +featureGate=DRAConsumableCapacity
+   Shared *bool `json:"shared,omitempty" protobuf:"bytes,7,opt,name=shared"`
```

Comment:
It possible to claim multiple time the same device. Would it make sense to add another key to the `DeviceRequestAllocationResult`? Like `SharedIndex` or `SharedUid`? I am asking because of KEP-4817 (Device Status). If a device is shared (let's say net1) and we create 2 MACVLANs on top of it (ethA and ethB), then they will both individual device status.

`SharedIndex` or `SharedUid` would be nil if device is not shared (so no need for `Shared *bool`)

Reply:
As discussed in the WG Device Management meeting, let's go with UID. We could generate a unique index because it only needs to be unique within the claim, but if we ever allow reallocation or resizing after allocation, then an index might not be stable enough.

Instead of `SharedUid` I would like to propose `ShareUID` because this is "the globally unique identifier of the share allocated from the shared device".

`SharedUID` sounds too much like "a shared UID" which doesn't make sense...